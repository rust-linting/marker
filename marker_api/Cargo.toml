[package]
name = "marker_api"

categories  = ["development-tools"]
description = "Marker's API, designed for stability and usability"

edition      = { workspace = true }
keywords     = { workspace = true }
license      = { workspace = true }
repository   = { workspace = true }
rust-version = { workspace = true }
version      = { workspace = true }

[dependencies]
typed-builder = { workspace = true, optional = true }
visibility    = { workspace = true, optional = true }

[dev-dependencies]
expect-test = { workspace = true }

[features]
# Some items should only be used by the driver implementing the functionality,
# this feature enables the export of these items. Note that this interface is
# unstable. All usage of the driver api can change between releases.
driver-api = ["dep:visibility", "dep:typed-builder"]
