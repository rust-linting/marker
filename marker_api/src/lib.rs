#![doc = include_str!("../README.md")]
#![warn(clippy::pedantic)]
#![allow(clippy::module_name_repetitions)]
#![allow(clippy::must_use_candidate)]
#![allow(clippy::trivially_copy_pass_by_ref)]
#![allow(clippy::unused_self)] // `self` is needed to change the behavior later
#![allow(clippy::missing_panics_doc)] // Temporary allow for `todo!`s
#![allow(clippy::new_without_default)] // Not very helpful as `new` is almost always cfged
#![allow(clippy::no_effect_underscore_binding)] // FP with derive macros clippy#12045
#![cfg_attr(not(feature = "driver-api"), allow(dead_code))]
#![cfg_attr(marker, warn(marker::marker_lints::not_using_has_span_trait))]

pub static MARKER_API_VERSION: &str = env!("CARGO_PKG_VERSION");

mod interface;
mod private;
pub use interface::*;
mod lint;
pub use lint::*;

#[cfg(test)]
pub(crate) mod test;

pub mod ast;
pub mod common;
pub mod context;
pub mod diagnostic;
pub mod prelude;
pub mod sem;
pub mod span;

#[doc(hidden)]
pub mod ffi;

pub use context::MarkerContext;
pub use interface::{LintPassInfo, LintPassInfoBuilder};

/// A [`LintPass`] visits every node like a `Visitor`. The difference is that a
/// [`LintPass`] provides some additional information about the implemented lints.
/// The adapter will walk through the entire AST once and give each node to the
/// registered [`LintPass`]es.
pub trait LintPass {
    fn info(&self) -> LintPassInfo;

    fn check_crate<'ast>(&mut self, _cx: &'ast MarkerContext<'ast>, _krate: &'ast ast::Crate<'ast>) {}
    fn check_item<'ast>(&mut self, _cx: &'ast MarkerContext<'ast>, _item: ast::ItemKind<'ast>) {}
    fn check_field<'ast>(&mut self, _cx: &'ast MarkerContext<'ast>, _field: &'ast ast::ItemField<'ast>) {}
    fn check_variant<'ast>(&mut self, _cx: &'ast MarkerContext<'ast>, _variant: &'ast ast::EnumVariant<'ast>) {}
    fn check_body<'ast>(&mut self, _cx: &'ast MarkerContext<'ast>, _body: &'ast ast::Body<'ast>) {}
    fn check_stmt<'ast>(&mut self, _cx: &'ast MarkerContext<'ast>, _stmt: ast::StmtKind<'ast>) {}
    fn check_expr<'ast>(&mut self, _cx: &'ast MarkerContext<'ast>, _expr: ast::ExprKind<'ast>) {}
}

/// This struct blocks the construction of enum variants, similar to the `#[non_exhaustive]`
/// attribute.
///
/// Marker uses enums extensively, like [`ItemKind`][ast::ItemKind] and
/// [`ExprKind`](ast::ExprKind). There can be `*Kind` enums that wrap other
/// `*Kind` enums. In those cases, this struct is used, to block the user from
/// constructing the variant manually. This allows tools to handle the variants
/// confidently without additional verification. An example for this would be the
/// [`LitExprKind::UnaryOp`](ast::LitExprKind::UnaryOp) variant.
///
/// This basically acts like a `#[non_exhaustive]` attribute, with the difference
/// that it also works on tuple variants. Attaching `#[non_exhaustive]` to a tuple
/// variant would make it private, which we don't want.
///
/// As a normal user, you can just ignore this instance as it holds no relevant
/// information for linting.
#[repr(C)]
#[non_exhaustive]
#[derive(Copy, Clone)]
pub struct CtorBlocker {
    /// `#[repr(C)]` requires a field, to make this a proper type. This is just
    /// the smallest one.
    _data: u8,
}

impl std::fmt::Debug for CtorBlocker {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("..").finish()
    }
}

impl CtorBlocker {
    #[cfg_attr(feature = "driver-api", visibility::make(pub))]
    pub(crate) fn new() -> Self {
        Self { _data: 255 }
    }
}
