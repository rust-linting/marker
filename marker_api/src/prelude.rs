//! This prelude is a collection of traits and types which are commonly used
//! when working with Marker. Simply add `use marker_api::prelude::*;` to your
//! file, to import them all.

// Traits:
pub use crate::ast::expr::ExprData;
pub use crate::ast::item::ItemData;
pub use crate::ast::pat::PatData;
pub use crate::ast::stmt::StmtData;
pub use crate::ast::ty::SynTyData;
pub use crate::ast::HasNodeId;
pub use crate::diagnostic::EmissionNode;
pub use crate::span::HasSpan;

// IDs
pub use crate::ast::{BodyId, ExprId, FieldId, GenericId, ItemId, NodeId, StmtId, TyDefId, VarId, VariantId};

// Common types
pub use crate::ast::expr::ExprKind;
pub use crate::ast::item::Body;
pub use crate::ast::item::ItemKind;
pub use crate::ast::pat::PatKind;
pub use crate::ast::stmt::StmtKind;
pub use crate::ast::ty::SemTyKind;
pub use crate::ast::ty::SynTyKind;
pub use crate::span::Ident;
pub use crate::span::Span;
pub use crate::MarkerContext;
