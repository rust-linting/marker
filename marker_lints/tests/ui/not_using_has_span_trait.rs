#![cfg_attr(marker, warn(marker::marker_lints::not_using_has_span_trait))]

use marker_api::prelude::*;

pub fn blackjack(_: u32, _: &Span<'_>) {}
pub fn rampage(_: &marker_api::span::Span<'_>, _: bool) {}

fn main() {}
