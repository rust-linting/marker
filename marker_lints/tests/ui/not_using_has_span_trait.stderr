warning: use impl HasSpan instead of Span for more flexibility
 --> $DIR/not_using_has_span_trait.rs:5:29
  |
5 | pub fn blackjack(_: u32, _: &Span<'_>) {}
  |                             ^^^^^^^^^
  |
note: the lint level is defined here
 --> $DIR/not_using_has_span_trait.rs:1:26
  |
1 | #![cfg_attr(marker, warn(marker::marker_lints::not_using_has_span_trait))]
  |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: use impl HasSpan instead of Span for more flexibility
 --> $DIR/not_using_has_span_trait.rs:6:19
  |
6 | pub fn rampage(_: &marker_api::span::Span<'_>, _: bool) {}
  |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: 2 warnings emitted

